### PR TITLE
Fix FP reported in #388

### LINKED ISSUE #388: `M0-1-4`: Recognise aggregate literal initializers
### Affected rules
 - `M0-1-4`

### Description

`M0-1-4` identifies single use variables, but for member variables we do not recognise the initialization of variables in an aggregate initializer as a use.

In addition, we should search for other places where we count uses of member variables to ensure this case is also captured there.

### Example

```cpp
struct agg {
  int x;
};

void test_agg() {
  agg a = {1};
  a.x;
}
```

### PR DESCRIPTION
**Pull request by rvermeulen**
## Description

Resolve #388 

## Change request type

 - [ ] Release or process automation (GitHub workflows, internal scripts)
 - [ ] Internal documentation
 - [ ] External documentation
 - [x] Query files (`.ql`, `.qll`, `.qls` or unit tests)
 - [ ] External scripts (analysis report or other code shipped as part of a release)

## Rules with added or modified queries

 - [ ] No rules added
 - [ ] Queries have been added for the following rules:
     - _rule number here_
 - [x] Queries have been modified for the following rules:
     - M0-1-4

## Release change checklist

A change note ([development_handbook.md#change-notes](https://github.com/github/codeql-coding-standards/blob/main/docs/development_handbook.md#change-notes)) is required for any pull request which modifies:

 - The structure or layout of the release artifacts.
 - The evaluation performance (memory, execution time) of an existing query.
 - The results of an existing query in any circumstance.

If you are only adding new rule queries, a change note is not required.

_**Author:**_ Is a change note required? 
  - [x] Yes
  - [ ] No

🚨🚨🚨
_**Reviewer:**_ Confirm that format of *shared* queries (not the .qll file, the
.ql file that imports it) is valid by running them within VS Code. 
  - [x] Confirmed

_**Reviewer:**_ Confirm that either a change note is not required or the change note is required and has been added.
  - [x] Confirmed

## Query development review checklist

For PRs that add new queries or modify existing queries, the following checklist should be completed by both the author and reviewer:

### Author

 - [x] Have all the relevant rule package description files been checked in?
 - [x] Have you verified that the metadata properties of each new query is set appropriately?
 - [x] Do all the unit tests contain both "COMPLIANT" and "NON_COMPLIANT" cases?
 - [x] Are the alert messages properly formatted and consistent with the [style guide](https://github.com/github/codeql-coding-standards/blob/main/docs/development_handbook.md#query-style-guide)?
 - [x] Have you run the queries on OpenPilot and verified that the performance and results are acceptable?<br />_As a rule of thumb, predicates specific to the query should take no more than 1 minute, and for simple queries be under 10 seconds. If this is not the case, this should be highlighted and agreed in the code review process._
 - [x] Does the query have an appropriate level of in-query comments/documentation?
 - [x] Have you considered/identified possible edge cases?
 - [x] Does the query not reinvent features in the standard library?
 - [x] Can the query be simplified further (not golfed!)

### Reviewer 

 - [x] Have all the relevant rule package description files been checked in?
 - [x] Have you verified that the metadata properties of each new query is set appropriately?
 - [x] Do all the unit tests contain both "COMPLIANT" and "NON_COMPLIANT" cases?
 - [x] Are the alert messages properly formatted and consistent with the [style guide](https://github.com/github/codeql-coding-standards/blob/main/docs/development_handbook.md#query-style-guide)?
 - [x] Have you run the queries on OpenPilot and verified that the performance and results are acceptable?<br />_As a rule of thumb, predicates specific to the query should take no more than 1 minute, and for simple queries be under 10 seconds. If this is not the case, this should be highlighted and agreed in the code review process._
 - [x] Does the query have an appropriate level of in-query comments/documentation?
 - [x] Have you considered/identified possible edge cases?
 - [x] Does the query not reinvent features in the standard library?
 - [x] Can the query be simplified further (not golfed!)

### CODE REVIEW

File: cpp/autosar/src/rules/M0-1-4/SingleUsePODVariable.qll [hunk 1]
```diff
@@ -10,23 +10,69 @@ private string getConstExprValue(Variable v) {
...
-        // For constexpr variables used as template arguments, we don't see accesses (just the
-        // appropriate literals). We therefore take a conservative approach and count the number of
-        // template instantiations that use the given constant, and consider each one to be a use
-        // of the variable
-        count(ClassTemplateInstantiation cti |
-          cti.getTemplateArgument(_).(Expr).getValue() = getConstExprValue(v)
-        )
-  )
+  // We enforce that it's a POD type variable, so if it has an initializer it is explicit
+  //v.getFile().getBaseName() = "test_member.cpp" and
```
Comment (knewbury01) [inline cpp/autosar/src/rules/M0-1-4/SingleUsePODVariable.qll:54]: oop, testing residual comment
```suggestion
```

File: cpp/autosar/src/rules/M0-1-4/SingleUseMemberPODVariable.ql [hunk 2]
```diff
@@ -24,5 +24,5 @@ where
   not isExcluded(v, DeadCodePackage::singleUseMemberPODVariableQuery()) and
   isSingleUseNonVolatilePODVariable(v)
 select v,
-  "Member POD variable " + v.getName() + " in " + v.getDeclaringType().getName() + " is only $@.",
-  getSingleUse(v), "used once"
+  "Member POD variable '" + v.getName() + "' in '" + v.getDeclaringType().getName() +
```
Comment (knewbury01) [inline cpp/autosar/src/rules/M0-1-4/SingleUseMemberPODVariable.ql:27]: not sure if while we are at it, we could apply this similar alert format improvement to the other queries in this rule

File: cpp/autosar/src/rules/M0-1-4/SingleUsePODVariable.qll [hunk 3]
```diff
@@ -10,23 +10,69 @@ private string getConstExprValue(Variable v) {
...
+/**
+ * Gets the number of uses of variable `v` in an opaque assignment, where an opaqua assignment for example a cast from one type to the other and `v` is assumed to be a member of the resulting type.
+ * e.g.,
+ * struct foo {
+ *  int bar;
+ * }
+ *
+ * struct foo * v = (struct foo*)buffer;
+ */
+Expr getIndirectSubObjectAssignedValue(MemberVariable subobject) {
```
Comment (knewbury01) [inline cpp/autosar/src/rules/M0-1-4/SingleUsePODVariable.qll:22]: without looking at this too closely (but just to double check)... is there anything in the existing [aggregate literal initializers concepts](https://github.com/github/codeql-coding-standards/blob/main/cpp/common/src/codingstandards/cpp/enhancements/AggregateLiteralEnhancements.qll) already that can be reused here?

**Review by knewbury01 — CHANGES_REQUESTED** [review]
just a few minor things/things to double check

File: cpp/autosar/src/rules/M0-1-4/SingleUsePODVariable.qll (hunk 3 above)
Comment (rvermeulen) [reply cpp/autosar/src/rules/M0-1-4/SingleUsePODVariable.qll:22]: No, this looks at cases where you for example cast a byte buffer to a struct. This reinterpretation "assigns" a value to a field of struct, without actually assigning it. However, we still want to count that as a use to ensure we alert on those fields.

File: cpp/autosar/src/rules/M0-1-4/SingleUseMemberPODVariable.ql (hunk 2 above)
Comment (rvermeulen) [reply cpp/autosar/src/rules/M0-1-4/SingleUseMemberPODVariable.ql:27]: We will address incorrectly formatted alerts in a different PR once we have QL4QL rule for it. For now I'll do it for queries that I'm changing anyways.

**Pull request merged**